### PR TITLE
Added Time Types: Date, Timestamp and Duration

### PR DESCRIPTION
**Pull request by L0Lmaker**
*Issue #, if available:*

https://bitquill.atlassian.net/jira/software/c/projects/AN/boards/36?modal=detail&selectedIssue=AN-1007

*Description of changes:*

Added Support for time related datatypes.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by lyndonbauto** [comment]
@L0Lmaker Can you redirect this pull request at gremlin-go-ms3? You will likely need to rebase.

File: gremlin-go/driver/graphBinary.go [hunk 1]
```diff
@@ -577,6 +581,53 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
...
+	if err != nil {
+		return nil, err
+	}
+	return time.UnixMilli(newTime.(int64)), nil
+}
+
+func durationWriter(value interface{}, buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) ([]byte, error) {
+	t := value.(time.Duration)
+	sec := int64(t.Seconds())
+	millis := t.Seconds() - float64(sec)
```
Comment (simonz-bq) [inline gremlin-go/driver/graphBinary.go]: Why did you convert Seconds, which is a float64 to int64, and then convert it bakc into float64, in order to subtract it from itself?

File: gremlin-go/driver/graphBinary.go [hunk 2]
```diff
@@ -577,6 +581,53 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
...
+	nanos := int32(millis * 1000000000)
```
Comment (simonz-bq) [inline gremlin-go/driver/graphBinary.go]: Can't this just be `t.Nanoseconds()`?

File: gremlin-go/driver/graphBinary.go [hunk 3]
```diff
@@ -577,6 +581,53 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
...
+func durationReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) (interface{}, error) {
+	sec, err := typeSerializer.read(buffer)
+	if err != nil {
+		return nil, err
+	}
+	nanosec, err := typeSerializer.read(buffer)
+	if err != nil {
+		return nil, err
+	}
+	total := sec.(int64)*int64(1000000) + int64(nanosec.(int32))
```
Comment (simonz-bq) [inline gremlin-go/driver/graphBinary.go]: Can this conversion be done using the time duration library?

**Review by simonz-bq — CHANGES_REQUESTED** [review]

File: gremlin-go/driver/graphBinary.go (hunk 1 above)
Comment (L0Lmaker) [reply gremlin-go/driver/graphBinary.go]: The format for Duration on Tinkerpop is `Format: {seconds}{nanos}`
I will add this as a comment before the writer function. The reason I did it like that was to convert pure nanoseconds to a combination of seconds + nanoseconds.

File: gremlin-go/driver/graphBinary.go (hunk 2 above)
Comment (L0Lmaker) [reply gremlin-go/driver/graphBinary.go]: Previous comment

**Review by simonz-bq — APPROVED** [review]

File: gremlin-go/driver/graphBinary.go [hunk 4]
```diff
@@ -579,6 +583,52 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
 	return p, nil
 }
 
+func timeWriter(value interface{}, buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) ([]byte, error) {
+	t := value.(time.Time)
+	_, err := typeSerializer.write(t.UnixMilli(), buffer)
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: writeValue

File: gremlin-go/driver/graphBinary.go [hunk 5]
```diff
@@ -579,6 +583,52 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
...
+	if err != nil {
+		return nil, err
+	}
+	return buffer.Bytes(), nil
+}
+
+func timeReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) (interface{}, error) {
+	newTime, err := typeSerializer.read(buffer)
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: readValue

File: gremlin-go/driver/graphBinary.go [hunk 6]
```diff
@@ -579,6 +583,52 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
...
+	}
+	_, err = typeSerializer.write(nanos, buffer)
+	if err != nil {
+		return nil, err
+	}
+	return buffer.Bytes(), nil
+}
+
+func durationReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer) (interface{}, error) {
+	sec, err := typeSerializer.read(buffer)
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: readValue

File: gremlin-go/driver/graphBinary.go [hunk 7]
```diff
@@ -579,6 +583,52 @@ func pathReader(buffer *bytes.Buffer, typeSerializer *graphBinaryTypeSerializer)
...
+	if err != nil {
+		return nil, err
+	}
+	nanosec, err := typeSerializer.read(buffer)
```
Comment (lyndonbauto) [inline gremlin-go/driver/graphBinary.go]: readValue

**Pull request merged**